pygdf/libgdf_groupby.py

```diff
@@ -0,0 +1,246 @@
+
+            result[val_columns_out[col_count]] = out_col_agg_series[:num_row_results]
+
+            out_col_agg_series.data.size = num_row_results
+            out_col_agg_series = out_col_agg_series.reset_index()
+
+            first_run = False
+            col_count = col_count + 1
+
+        return result
+
+    def min(self):
+        agg_type = "min"
+
+        result = DataFrame()
+        add_col_values = True
+
+        ctx = ffi.new('gdf_context*')
+        ctx.flag_sorted = 0
+        ctx.flag_method = self._method
+        ctx.flag_distinct = 0
+
+        val_columns = self._val_columns
+
+        return self._apply_agg(
+            agg_type, result, add_col_values, ctx, val_columns)
+
+    def max(self):
+        agg_type = "max"
+
+        result = DataFrame()
+        add_col_values = True
+
+        ctx = ffi.new('gdf_context*')
+        ctx.flag_sorted = 0
+        ctx.flag_method = self._method
+        ctx.flag_distinct = 0
+
+        val_columns = self._val_columns
+
+        return self._apply_agg(
+            agg_type, result, add_col_values, ctx, val_columns)
+
+    def count(self):
+        agg_type = "count"
+
+        result = DataFrame()
+        add_col_values = True
+
+        ctx = ffi.new('gdf_context*')
+        ctx.flag_sorted = 0
+        ctx.flag_method = self._method
+        ctx.flag_distinct = 0
+
+        val_columns = self._val_columns
+
+        return self._apply_agg(
+            agg_type, result, add_col_values, ctx, val_columns)
+
+    def sum(self):
+        agg_type = "sum"
+
+        result = DataFrame()
+        add_col_values = True
+
+        ctx = ffi.new('gdf_context*')
+        ctx.flag_sorted = 0
+        ctx.flag_method = self._method
+        ctx.flag_distinct = 0
+
+        val_columns = self._val_columns
+
+        return self._apply_agg(
+            agg_type, result, add_col_values, ctx, val_columns)
+
+    def mean(self):
+        agg_type = "mean"
+
+        result = DataFrame()
+        add_col_values = True
+
+        ctx = ffi.new('gdf_context*')
+        ctx.flag_sorted = 0
+        ctx.flag_method = self._method
+        ctx.flag_distinct = 0
+
+        val_columns = self._val_columns
+
+        return self._apply_agg(
+            agg_type, result, add_col_values, ctx, val_columns)
+
+    def agg(self, args):
+
+        result = DataFrame()
+        add_col_values = True
+
+        ctx = ffi.new('gdf_context*')
+        ctx.flag_sorted = 0
+        ctx.flag_method = self._method
+        ctx.flag_distinct = 0
+
+        if isinstance(args, (tuple, list)):
+            for agg_type in args:
+
+                val_columns_out = [val + '_' +
+                                   agg_type for val in self._val_columns]
+
+                result = self._apply_agg(
+                    agg_type, result, add_col_values, ctx, self._val_columns, val_columns_out)
+
+                add_col_values = False  # we only want to add them once
+
+        elif isinstance(args, dict):
```

Comment:
likewise, this could be the more general `elif isinstance(args, collections.abc.Mapping)`